ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Impl/Validation.hs

```diff
@@ -0,0 +1,414 @@
+    let (epochFiles, _, _) = dbFilesOnDisk filesInDBFolder
+    case Set.lookupMax epochFiles of
+      Nothing              -> do
+        -- Remove left-over index files
+        -- TODO calls listDirectory again
+        removeFilesStartingFrom hasFS 0
+        return (0, TipGen)
+
+      Just lastEpochOnDisk -> case valPol of
+        ValidateAllEpochs       ->
+          validateAllEpochs       validateEnv lastEpochOnDisk
+        ValidateMostRecentEpoch ->
+          validateMostRecentEpoch validateEnv lastEpochOnDisk
+  where
+    HasFS { listDirectory } = hasFS
+
+-- | Validate epochs from oldest to newest, stop at the most recent epoch on
+-- disk. During this validation, keep track of the last valid block in each
+-- epoch. If at the end, that block is not in the last epoch on disk, remove
+-- the epoch and index files after that epoch.
+--
+-- TODO stop when a block is missing, detect this by checking whether the
+-- hashes line up
+validateAllEpochs
+  :: forall m hash h e. (IOLike m, Eq hash, HasCallStack)
+  => ValidateEnv m hash h e
+  -> EpochNo
+     -- ^ Most recent epoch on disk
+  -> m (EpochNo, ImmTipWithHash hash)
+validateAllEpochs validateEnv@ValidateEnv { hasFS, err, epochInfo } lastEpoch =
+    go (0, TipGen) 0 Origin
+  where
+    go
+      :: (EpochNo, ImmTipWithHash hash)  -- ^ The last valid epoch and tip
+      -> EpochNo                         -- ^ The epoch to validate now
+      -> WithOrigin hash                 -- ^ The hash of the last block of
+                                         -- the previous epoch
+      -> m (EpochNo, ImmTipWithHash hash)
+    go lastValid epoch prevHash = do
+      shouldBeFinalised <- if epoch == lastEpoch
+        then return ShouldNotBeFinalised
+        else ShouldBeFinalised <$> epochInfoSize epochInfo epoch
+      runExceptT
+        (validateEpoch validateEnv shouldBeFinalised epoch (Just prevHash)) >>= \case
+          Left  ()              -> cleanup lastValid epoch $> lastValid
+          Right Nothing         -> continueOrStop lastValid             epoch prevHash
+          Right (Just validBlk) -> continueOrStop (epoch, Tip validBlk) epoch prevHash'
+            where
+              prevHash' = At (snd validBlk)
+
+    -- | Validate the next epoch, unless the epoch just validated is the last
+    -- epoch to validate. Cleanup files corresponding to epochs after the
+    -- epoch in which we found the last valid block. Return that epoch and the
+    -- tip corresponding to that block.
+    continueOrStop
+      :: (EpochNo, ImmTipWithHash hash)
+      -> EpochNo         -- ^ The epoch just validated
+      -> WithOrigin hash -- ^ The hash of the last block of the previous epoch
+      -> m (EpochNo, ImmTipWithHash hash)
+    continueOrStop lastValid epoch prevHash
+      | epoch < lastEpoch
+      = go lastValid (epoch + 1) prevHash
+      | otherwise
+      = assert (epoch == lastEpoch) $ do
+        cleanup lastValid epoch
```

Comment:
Add a comment: cleanup is only needed when the final epoch was empty, yet valid (L129). So we need to remove it and unfinalise it.